test/integration/idle_timeout_integration_test.cc

```diff
@@ -178,8 +193,94 @@ TEST_P(IdleTimeoutIntegrationTest, PerStreamIdleTimeoutAfterBidiData) {
+  testRouterHeaderOnlyRequestAndResponse(true, nullptr);
+}
+
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutConfiguredRequestResponseWithBody) {
+  enable_request_timeout_ = true;
   testRouterRequestAndResponseWithBody(1024, 1024, false, nullptr);
 }
 
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutTriggersOnBodilessPost) {
+  enable_request_timeout_ = true;
+
+  auto response = setupPerStreamIdleTimeoutTest("POST");
+
+  waitForTimeout(*response, "downstream_rq_timeout");
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+  EXPECT_TRUE(response->complete());
+  EXPECT_STREQ("408", response->headers().Status()->value().c_str());
+  EXPECT_EQ("request timeout", response->body());
+}
+
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutUnconfiguredDoesNotTriggerOnBodilessPost) {
+  enable_request_timeout_ = false;
+  // with no request timeout configured, the idle timeout triggers instead
+  enable_per_stream_idle_timeout_ = true;
+
+  auto response = setupPerStreamIdleTimeoutTest("POST");
+
+  waitForTimeout(*response);
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+  EXPECT_TRUE(response->complete());
+  EXPECT_STREQ("408", response->headers().Status()->value().c_str());
+  EXPECT_NE("request timeout", response->body());
+}
+
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutTriggersOnRawIncompleteRequestWithHeaders) {
+  if (downstreamProtocol() == Envoy::Http::CodecClient::Type::HTTP2) {
+    return;
+  }
+  enable_request_timeout_ = true;
+
+  initialize();
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+  std::string raw_response;
+  sendRawHttpAndWaitForResponse(lookupPort("http"), "GET / HTTP/1.1", &raw_response, true);
+  EXPECT_THAT(raw_response, testing::HasSubstr("request timeout"));
+}
+
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutDoesNotTriggerOnRawCompleteRequestWithHeaders) {
+  if (downstreamProtocol() == Envoy::Http::CodecClient::Type::HTTP2) {
+    return;
+  }
+  enable_request_timeout_ = true;
+
+  initialize();
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+  std::string raw_response;
+  sendRawHttpAndWaitForResponse(lookupPort("http"), "GET / HTTP/1.1\r\n\n\n", &raw_response, true);
+  EXPECT_THAT(raw_response, testing::Not(testing::HasSubstr("request timeout")));
+}
+
+TEST_P(IdleTimeoutIntegrationTest, RequestTimeoutIsDisarmedByPrematureEncodeHeaders) {
+  enable_request_timeout_ = true;
+  enable_per_stream_idle_timeout_ = true;
+
+  auto response = setupPerStreamIdleTimeoutTest("POST");
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+
+  waitForTimeout(*response);
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_FALSE(response->complete());
+  EXPECT_NE("request timeout", response->body());
+}
+
+// TODO(auni53) create a test filter that hangs and does not send data upstream, which would
```

Comment:
Are you planning to do a followup PR with this?  If so, want to file an issue and assign it to yourself?

Reply:
ummm, let me actually work on this now. If I get a LGTM on this PR while it's wip I'll create an issue.

Reply:
nvm I don't want to do this now, added an issue #4988